### DISCUSSION TITLE
Add "virtual private" accessibility mode

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/csharplang#5885] [Language Ideas]
Currently, there are two reasons for exposing a base method to derivative classes as "protected":

1. You want derivatives to be able to see or call this method.
2. OR, you simply want derivatives to be able to OVERRIDE, but NOT CALL this method.

Given that C# is currently unable to specify "usage" on "virtual" methods, for sake of QA safety, many developers have adopted this pattern:

```c#
public void Update()
{
   // do some stuff

   OnUpdate();

   // do more stuff
}
protected virtual void OnUpdate() { }  // method should ONLY be called from "Update" method
```

So this introduces two issues for derivative classes:
1. Now Intellisense is constantly showing them "OnUpdate" as a viable method that can be called -- when it should NOT ever be called by derivative classes.  OnUpdate should ONLY be called from "Update()" method.
2. If they call "OnUpdate()" defects are likely to occur.

I am proposing that we introduce into the C# language the combo term **"virtual private"**, which means that "you can override, but not call from derived classes".

The above code would become:
```c#
public void Update()
{
   // do some stuff

   OnUpdate();

   // do more stuff
}
private virtual void OnUpdate() { }  // method CAN ONLY BE OVERRIDDEN, not called from derived classes
```

This then resolves the two issues of "cluttering intellisense" as well as "preventing defects of this nature".

**Comment by HaloFour** [comment]
> Now Intellisense is constantly showing them "OnUpdate" as a viable method that can be called

So inform tooling to not show the method:

```cs
[EditorBrowsable(EditorBrowsableState.Never)]
protected virtual void OnUpdate() { }
```

A method is either accessible or it's not.  If it's accessible, it's invokable.

**Reply by najak3d** [reply]
The [System.ComponentModel.EditorBrowsableState] attribute only works within the same assembly, and so is partially deficient, plus is a lot to type.

Your absolute statement about "it's accessible or it's not" may be "how it is now" but doesn't need to be that way.

For example, properties marked "init" or Fields marked "readonly" -- are READONLY.    So the current situation isn't that simple for all things.   There are also smart suggestions on this site for "init methods" for methods that can only be called during initialization process.

So, IMO, that's a weak argument to say "just because you can override a method, also means that you MUST be able to also CALL it".  Why argue for that rule?

"private virtual" - makes intuitive sense.    "virtual" mean overridable.   "private" means hidden from derivatives.      Add those together and you get an accessor that means "overridable but not otherwise visible to the derivative classes".

**Reply by HaloFour** [reply]
> and so is partially deficient.

You keep using that word. I do not think it means what you think it means.

> "private virtual" - makes intuitive sense.

If it's inaccessible then by definition it's not overridable.  It's cannot be overridden if it's inaccessible.

> That attribute only works within the same assembly

Again, analyzers were added to the compiler _very specifically_ to enable _you_ to define the guardrails you want to impose, because the language is not going to add new keywords for every possible niche scenario that arises.

**Reply by najak3d** [reply]
In this case, I'm not asking for a "new keyword" but simply to allow a notation that actually makes perfect sense.

What else could "private virtual" mean?  For methods, 'private' controls "who can call it".   "virtual" controls whether or not you can override.  These terms should NOT be related.   They are entirely separate concerns/concepts.

"private virtual"  -- should be valid.  It's clear, and useful.  And an easy upgrade for C# to make.

**Reply by najak3d** [reply]
"Deficient" means "lacking".   Compared to my proposal for "private virtual", the use of that the [System.ComponentModel.EditorBrowsableState] attribute is deficient in two ways:

1. Terseness.   It's verbose.
2. Functionality.   It only limits visibility outside of your assembly

So deficient seems to be an appropriate word here.   The attribute is deficient when compared to the proposed solution.

**Reply by HaloFour** [reply]
@najak3d 

It's new syntax, regardless of whether it's trying to repurpose existing keywords.  It's not a goal of the language to solve every single problem that arises through syntax.

> For methods, 'private' controls "who can call it". "virtual" controls whether or not you can override. These terms should NOT be related. They are entirely separate concerns/concepts.

They are already related.  You cannot override `internal virtual` methods from outside of the assembly, because if the method is not accessible, it's also not overridable.

> Terseness. It's verbose.

Uncommon scenarios don't necessitate terse solutions.

> Functionality. It only limits visibility outside of your assembly

Take that up with the tooling team, the C# language doesn't dictate how autocomplete happens to work.  I also think analyzers are capable of affecting autocomplete in situations like this.

**Reply by najak3d** [reply]
You raising the point of "internal virtual" is indicates yet another deficiency of C#.  So if you want to limit the override to within an Assembly, now you MUST also make it callable by ANY class within the assembly!

This is a silly relationship, between "who can call a method" vs. "who can override that method"... these are completely different concerns here.

If they want to control scope of "who can override" then that should be done separately from "who can call it". 

For this weird case of "internal virtual" -- this needs to be augmented to be sane!...  I'd suggest allowing this notation for this very specialized case:

protected virtual internal void MethodName() {}


The rule for that would be simple -- for virtual methods, and there **can be** two accessibility keywords, the one that comes after 'virtual' would be the one that says "who can override" ... in this case, the method can be called only by derivatives (even outside the assembly), but cannot be overridden from outside the assembly.

The absence of the 2nd accessibility modifier defaults it's override accessibility to the first accessibility modifier (where private implies protected).

Accessibility of "callers" vs. "overrides" are fully separate concerns, and should never have been intertwined.

**Reply by HaloFour** [reply]
> You raising the point of "internal virtual" is indicates yet another deficiency of C#

And of every other OOP language that exists?  A member is either accessible or it's not.

> protected virtual internal void MethodName() {}

That syntax is already legal.  C# already allows modifiers to appear in any order.

**Reply by najak3d** [reply]
"protected virtual internal" is not already legal.  Compiler error every time.   But it *should* be legal, so that we can exert appropriate, sometimes separate accessibility of the Callers from the Overrides.

If we could rewind time to the invention of C#, but with advantage of foresight, would you really be voting to intertwine accessibility of Caller vs. Override?   Wouldn't you view these as separate concerns, and not really related at all.

How to Call, vs. How to Override -- two TOTALLY different concerns.    It was a mistake to intertwine them from the onset.

virtual private --  is clear and should be fine.
internal virtual protected -- should mean "internal for calling" and "protected for overrides".
protected virtual internal  -- should mean "callable from derivatives", and "can only be overridden from inside the assembly"

This is Simple, clear, and appropriate notation.

**Reply by HaloFour** [reply]
> "protected virtual internal" is not already legal

It compiles just fine:

https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwLABQAAgMwAERATGQMJkDehZzZADgE4D2CApgMa90ZAG5R2CAK4QUZWL3YxpsmAjIBZABQBKMgF4AfGQCMlEgG5CAXyA=

You may have an analyzer (e.g. R#) which will complain for stylistic reasons, but the compiler doesn't care.

**Reply by najak3d** [reply]
"protected virtual internal"  -  Ah crap - missed that.  I was so confident, I didn't even test it!..  :)

You never addressed my question about "what if we could rewind time?"  Which side of this argument would you be on then?  Would you vote for intertwining override accessibility with calling accessibility, or would you vote to keep them "intertwined by default, but allow for separate accessibility specifiers"... ??

**Reply by HaloFour** [reply]
I don't personally dwell on such hypotheticals because they can't really inform the thinking around the design we do have today.  I can imagine that a _lot_ would be different about .NET and C# if it were being designed from scratch today, perhaps to the point of being unrecognizable.  All that said, I still think that it'd be a tough sell that accessibility should be so finely-grained, that the benefit doesn't warrant the complexity.

**Reply by CyrusNajmabadi** [reply]
If it's private I would assume it could not even be seen, and would be fine to remove without a breaking change to other assemblies. This new semantics means it would both appear to be private (e.g. not visible outside the assembly), but also visible outside the assembly as part of your abi.

**Comment by HaloFour** [comment]
_replying to a deleted comment_:

Don't take it personally.  The vast majority of language proposals on this repository are very unlikely to get implemented, and that includes proposals suggested by the language design team themselves.  Language design is exceptionally hard, especially with a 20 year old language and billions of lines of existing code, and the language team is intentionally taking a very conservative approach.  Even with the features they'd like to see implemented they have a backlog that extends for probably more than a decade.  This is why the team has very intentionally attempted to build in ways to enable developers to solve their own problems via the compiler's APIs like analyzers and source generators.

Even if someone from the language team does decide to champion this issue that's only the first step in a long process.  Most language features take well over a year to hammer out and implement.  `!!`, despite being painfully simple and with zero runtime considerations, was originally proposed in 2019.

I think I've opened about a dozen proposals since Roslyn was first open-sourced in 2014.  I've only had one considered, and even that one had to go through a rigorous process.  It's not you.  😁

**Reply by HaloFour** [reply]
I will say that the superlatives and hyperbole don't really do much for the argument, though.  Claiming that C# is broken for every feature it doesn't have doesn't help advance your case.  Nobody will claim that C# is a perfect language, and everyone has very different (and often diametrically opposed) opinions as to what will improve the language.

**Reply by najak3d** [reply]
It's not that I expect it to be "implemented", because I know that's a long road.   Maneuvering C# is like altering course of a very large aircraft carrier.   Very costly and time consuming.

But I would expect good ideas to receive "more support".   I'm hearing quite a few nonsensical rebuttals against my arguments - which makes me think that these are fueled by some sort of bias.   I would expect more support for good ideas, rather than flat out rebuttals.

**Reply by HaloFour** [reply]
@najak3d 

> But I would expect good ideas to receive "more support".

Very few proposals on this repo are objectively "bad" ideas.  Most are "good" ideas that just don't bubble up to being worth the _massive_ time and effort that making a change to the language entails.

> I'm hearing quite a few nonsensical rebuttals against my arguments - which makes me think that these are fueled by some sort of bias

As suggested by @CyrusNajmabadi in the other discussion, you may also want to examine your biases.  In 8 years there's been zero feedback from others on this subject which is a pretty good indicator that such features do not have wide utility, especially compared to the many, many other proposals that would benefit a wide audience.

**Reply by najak3d** [reply]
I'm not speaking on bias here.  These are things that I've fought with for years... but NEVER thought before "how should C# do this differently?"   I just dealt with it as is, like everyone else.

Now that I'm here on this site, and I encounter issues as I program, I'm now actively asking myself, "isn't there a better way?"  So I have on a different mindset.  This is not bias; it's just an outside-the-box mindset.   What specific type of bias do you think I have here?

I guess you could say my bias is "optimistic" due to my newness to this topic of "improving C#" -- so I'll tend to favor changes;  while for the veterans here, you realize the costs/risks of each change, and so are biased towards realistic pessimism.   Obviously, I have to weigh heavily your experienced views more than my inexperienced views.

**Reply by CyrusNajmabadi** [reply]
>  What specific type of bias do you think I have here?

The biases I'm seeing are:

1. when you continuously insist that all the problems you are encountering are extremely important and relevant to fix for the entire ecosystem. 
2. When you steadfastly refuse any alternative already possible as being unacceptable based on your personal beliefs on what the ecosystem needs. 

This unwillingness to allow for any path other than the one you want is problematic and will likely lead to a lot of dissatisfaction on your part since all our proposals are built off of compromise and balancing many different needs and views.

**Reply by HaloFour** [reply]
_replying to several deleted comments_

Everyone is welcome to their opinion as to how the language is working for them.  Everyone has their pain points with the language.  Nobody will ever claim that C#, or any language is perfect.

But the design of C# can't be driven by the individual opinion.  It is driven by the aggregate of feedback here and through other channels.  Every change is incredibly expensive, especially for a 20 year old language with millions of active developers maintaining billions of lines of existing code.

But in my opinion C# is one of the more egalitarian languages.  The fact that we can even have this conversation in this forum I believe is evidence of that.  There are also community members, from many different backgrounds, who are actively contributing to the process in the form of specs, documentation and code.

**Comment by CyrusNajmabadi** [comment]
> I am proposing that we introduce into the C# language the combo term "virtual private", which means that "you can override, but not call from derived classes".

I would like to address this point on a meta level.  We get continual requests to take the language as is, and to *add* restrictive features to prevent users from a situation they don't like that the language allows.

We took a philosophical stance here that such a chain of requests is unending, and we don't want to be constantly in the game of adding restrictions.  After all, restrictions don't enable anyone to do anything new.  They only prevent people from doing something that some subset finds unpalatable.  We would instead prefer to work on *enabling* scenarios as when those scenarios are not possible, there's nothing anyone can do about that.

However, we are cognizant of the value in restriction, especially as it leads to domain-specific safety beneifts.   To that end we designed and nurtured an system to allow *anyone* to layer on additional analysis to the language that can provide said guardrails.  It is reasonably easy to both add such analysis and to distribute those analyzers to your own teams or to the broader C# ecosystem.  This frees the team from incredibly costly work to add niche features that do not enable, but only disable.  And it allow every part of the ecosystem to add rules that htey care about, no matter how niche they are.  Broadly valuable rules also can be picked up and easily and widely used by others in the ecosystem.

Philosophically, we are very happy with this arrangement.  And it is highly unlikely that continuing to insist that your set of niche rules is somehow important enough to outweigh that.  As others have said here, the restrictions you are asking for are things that practically no one has been asking for over the lifetime of C#, *esp* compared to many other areas that we get far more vociferous feedback on.  Given that even *those* much more engaged topics are still unlikely to see work on our part to make first class, i think you should expect that these much narrower and less valuable topics are even less likely than that to happen.

I'm also really glad you're posting here demonstrating your passion and desires to see C# become a better language for *you* (and the set of customer you think overlap with you).  I really hope the feedback you get doesn't dissuade you from continuing to give engage here.  We very much appreciate all the thought, effort and passion you've put into these posts.  However, we have to be clear that in no way will that necessarily translate to actual change happening that you want.  We will always be gauging and investing in the areas we truly believe serve to provide the most value *overall* to the ecosystem, and if that doesnt' align with the items you care deeply about, then that may just be life.  I truly hoe that if that happens that is not terribly discouraging for you and that you still see merit and purpose in conversing and engaging here.

Finally, you might want to consider bringing some of these ideas to discord first.  It appears that you do prefer a bit more of the synchronous back and forth style of bouncing ideas off of people.  You're absolutely welcome to continue that here, but you may find discord more amenable to your discussion style.

Thanks!

**Comment by CyrusNajmabadi** [comment]
> I'm getting the feeling that no matter how smart/good of a suggestion I make, the same few guys will shoot it down. I'm thinking it may be more of a matter of "who said it", rather than "what is said".

This is not true, and i feel like it's really necessary to nip this in the bud.  Not only is this style of rigorous analysis applied to all community posts, it's also applied to the posts taht the LDM members make themselves.  I myself get my own direct proposals shot down, and i've been involved with the language for 20 years. 

The problem i've seen here is that we have been trying to give you insight on how these language areas are thought about, but htat you have continually simply outright rejected that style of thinking as being unacceptable to you.  At that point if you're not willing ot consider or accept these things that are important to us, we will be at an impasse.

**Comment by TahirAhmadov** [comment]
To answer the feedback item of "nobody has ever requested this before", sometimes great ideas just come up in somebody's mind. Yes it's possible nobody thought of this before, but the answer to that is, better late than never, and also, I am a lot more inclined to congratulate the person who had the idea 20 years later than to scorn him, as opposed to everybody else who never had this idea.
WRT methods having to be visible to overriding and invoking at the same time, it's true that's how the language (in fact, most languages) work now. However, the more I think about it, the more I like the idea of de-coupling these two aspects. This concept is not unheard of. For example, we already have properties with differing getter/setter access levels. 
I do have to point out that this is a very involved change most likely, with questions such as, what changes will the CLR require, what to do with other members (like properties), is it possible to call the base from the overridden method, etc.
Regarding suggested workarounds, such as the attribute or analyzers: the attribute does not generate a compilation error, and even in VS you can configure it to ignore the attribute; and with analyzers, the problem is that the team developing a library already knows not to call that method from a child class, so it doesn't really add much value to guard themselves using an analyzer, whereas the consumers of the said library are under no obligation or expectation to use that analyzer, which may lead to incorrect usage of the lib and bugs.
Regarding the "enabling scenarios" vs "restricting" question, I agree and disagree. Yes we want to enable good scenarios, but we also want to disable bad scenarios. Importantly, this proposal does not hinder any current code. And I think the type of developer who looks for ways to introduce restrictions understands the immense difficulty of writing truly high quality code, as opposed to the mediocre mass who "just want to get stuff done". I'm not sure I would say I value restrictions **more** than "enablers", but definitely I value those equally.

**Reply by TahirAhmadov** [reply]
PS. Finally, how would [Obsolete("Don't call this!", true)] work in this case? Perhaps that's worth a try.

**Reply by najak3d** [reply]
IMO - I'd call that a hack, even if it worked, due to the confusing name.  Therefore it could never become a good standard solution, even if it works.   "private virtual" methods are not "obsolete".

**Reply by TahirAhmadov** [reply]
I'm not saying it's the solution, but if it works, it gets you 90% there - basically, it's just a style difference at that point. Give it a shot - I'm curious myself, because I would like to have this in some parts of my code, as well.

**Reply by najak3d** [reply]
Just tried it, in VS 2022 and the results aren't good.   It doesn't show my "message" in intellisense, just shows "deprecated" without the message -- so communicates "don't override or call this method" without that message showing.  Here's how it looks in VS2022.

![image](https://user-images.githubusercontent.com/10914515/156910234-715e544d-4923-4529-86cc-a3daf8210928.png)

**Comment by CyrusNajmabadi** [comment]
> and even in VS you can configure it to ignore the attribute; 

You will never get absolute safety in any features of these types. Reflection is always a thing. You will always have to accept users acting escape your protections.

**Reply by najak3d** [reply]
@CyrusNajmabadi  - NONE of this is about preventing hacking.  Everyone knows that when you use reflection to use an API in an unauthorized fashion - that QA is no longer guaranteed.   They ALREADY can do this with all private's.  So not sure why you are using this line of arguing here.

So all we're looking for here is assurance that MOST users will have prescriptions on usage.... so that if they want to break this prescription, they are going out of their way to do so (e.g. Reflection).

**Reply by CyrusNajmabadi** [reply]
>  So not sure why you are using this line of arguing here.

As stated, because the claim was:

> and even in VS you can configure it to ignore the attribute;

If the user is working to avoid guardrails, then you can't prevent that.

**Reply by TahirAhmadov** [reply]
@CyrusNajmabadi To be fair, that attribute is a much weaker guardrail than having to use reflection. But I know what you mean.

**Reply by CyrusNajmabadi** [reply]
@TahirAhmadov sure.  but the poitn was that someone was rejecting analyzers because hte user might go out of their way to suppress the intent of the library shipping with the analyzer.  Once your users are suppressing your intents, then that's no longer interesting to me as you've done your due diligence, and now you're at the point that you're just trying to enforce something with a disinterested party.  If you want a solution to that, then lang changes aren't the way to go either as you'll still find the enforcements non viable.

**Comment by bernd5** [comment]
In C++ private virtual exists which means exactly what the OP is proposing.

**Comment by najak3d** [comment]
If the only thing C# did was make "private virtual" work, that would satisfy 90% of the potential needs here.   It simply differentiates between reasons for exposing it to the derived classes -- is it for calling, or just overriding?   Private virtual means "override only" and you can only call the base method from inside the body of the override.   End of story.  Easy.  Useful.   Simple.  Aligns to C++.

**Reply by CyrusNajmabadi** [reply]
> Aligns to C++.

This is not a virtue.  We have no goal or general interest in aligning with c++.

**Reply by CyrusNajmabadi** [reply]
> If the only thing C# did was make "private virtual" work, that would satisfy 90% of the potential needs here

As mentioned before though, Analyzers also likely solve 90% of the potential need or more.  And analyzers already exist, work, and aligns with the motivations and goals i outlined above for this space of problems.  Repeating that you still want a language feature here despite having alternative solutions doesn't change the calculus of things.
